### ISSUE TITLE
Consider upgrading trace memoization failure to error

### ISSUE THREAD

**Issue by elliottslaughter** [StanfordLegion/legion#1042]
When a trace fails to be memoized, you'll get a warning like this:

```
[1 - 2aaaaac91800]    9.080019 {4}{runtime}: [warning 1097] LEGION WARNING: WARNING: The runtime has failed to memoize the trace more than 5 times, due to the absence of a replayable template. It is highly likely that trace 1 will not be memoized for the rest of execution. The most recent template was not replayable for the following reason: blocking call. Please change the mapper to stop making memoization requests.
 (from file /users/eslaught/regent-index-launch-sc21-oldeqcr/runtime/legion/legion_trace.cc:2010)
```

It's easy to miss this warning, particularly if your application spews a lot of uninitialized field warnings. And if you do you can see some pretty big runtime performance regressions.

I wonder if we should make this an error? It could be optional, but I think there's even an argument to be made for not doing that:

 1. The user has to explicitly opt-in to tracing by adding `begin_trace` and `end_trace` calls to the application code (or `__demand(__trace)` in Regent.
 2. The user further has to explicitly opt-in to memoization of the trace by setting it in the mapper, or by passing `-dm:memoize` to mappers based on the default mapper.
 3. There are a variety of flags that would disable this, like `-lg:no_physical_tracing` that would also provide ways to turn this off.

Since this is a performance issue, and not a correctness issue, and since the user basically has no reason to want memoization if it's not working, I think we should just make it a hard error if it fails. The only real reason I can see against this is that it _plausible_ that there might exist an application that fails 5 times but succeeds on the 6th. However, I'm not inclined to weigh this very heavily because I don't think we've ever actually seen an application that behaves like this. If we did, we could always add a flag that controls how many times a trace fails before the runtime gives up on it, or we could provide an opt-out option to turn it back into a warning.

This has recently affected two applications, S3D and experiments for SC21, and probably has others. I think there would be a substantial benefit to making this fail explicitly so that users can't ignore this issue in their code.

**Comment by lightsighter** [comment]
Assigning to @alexaiken so it can be discussed at a future meeting.

**Comment by lightsighter** [comment]
> since the user basically has no reason to want memoization if it's not working

I disagree with this. Users should be able to put `begin_trace` and `end_trace` around calls around a program that looks like this:

```
import libA
import libB
import libC

for i in iters:
    begin_trace()
    libA.method()
    libB.method()
    libC.method()
    end_trace()
```

Each of those libraries have their own mapper which will make different decisions about mapping and memoization. A failure to trace shouldn't be an error just because the user optimistically tried to get the libraries to work together.

I think it would be fine though to add a command line flag to the runtime like `-Wall` to promote all warnings to errors. I think we might already have that. I need to check.

**Comment by elliottslaughter** [comment]
If memoization is a composable and all three libraries opt in to memoization, I don't see why it shouldn't be a hard error if it fails.

Composability gives us the ability to say that if any three libraries can be memoized individually, they should also be able to memoized in any arbitrary composition. With that level of predictability, I think it's perfectly reasonable to say that we should let libraries opt in to memoization, take the `AND` of all libraries in a trace to see if memoization should be enabled overall, and issue a hard error if it doesn't work. If there is a bug in a library and it turns out it shouldn't have turned on memoization when it did, then there are flags to work around that.

I know memoization isn't actually composable right now, but arguably it should be, precisely because it's so valuable to have this sort of predictability. Even without a hard error, a lack of composability will mean that users will hit performance cliffs that they aren't in a position to resolve on their own, and I don't think we want that either.

**Comment by lightsighter** [comment]
Well, if @alexaiken gets his way with the changes that he already has proposed (I told him and @syamajala that they need to make a github issue for the changes, but they haven't done so yet), then this whole conversation is going to become mute because tracing will never fail again. It will just capture a non-idempotent trace and the runtime will issue the necessary copies at the replay point to bring the data up to date, so there will be a more graceful, but still present, degradation of performance. I still thinking having code run without failing should be the default. If users want to opt-in to stricter performance guard rails that is fine, but that is a decision for them to make. We shouldn't opt-in to the guard rails for them. If you want to be super precise we can even add a parameter to `begin_trace` to say what the setting is, but it should default off.

**Comment by lightsighter** [comment]
@elliottslaughter any objection to closing this issue now that we support non-idempotent traces?

**Comment by elliottslaughter** [comment]
Before we close this, can you enumerate the remaining patterns (if any) that would cause trace replay to fail?

**Comment by lightsighter** [comment]
There are two reasons that a trace replay will fail:

1. We cannot find a previously captured template for trace that has its preconditions satisfied by the currently valid region data. This is most likely because we've just never captured a template that matches the locations of the current valid data. Alternatively, you might have captured a template previously that would have had its preconditions satisfied by the current valid data, but the template was evicted from the cache of templates due to capacity reasons. You can control the size of this cache in the `configure_context` mapper call with the `max_templates_per_trace` parameter.
2. The other reason that a trace replay will fail is that we find a template with its preconditions met, but the template has internal instances that have been garbage collected (these would obviously be instances necessary for the template replay that are not part of the preconditions). After verifying the preconditions for a template, Legion tries to acquire references on the internal instances necessary for replay and if any of them fail then Legion evicts the template from the cache and proceeds to see if it can find another template whose preconditions can be satisfied.

I think this logic is more straightforward and ensures that as long as you have some kind of repeated mapping then you're likely to get replays regularly (unless you're very memory constrained and instances keep getting garbage collected).

**Comment by elliottslaughter** [comment]
Ok, this seems fair to me. Closing.
